src/block_service/hdfs/hdfs_service.cpp

```diff
@@ -0,0 +1,471 @@
+        if (hdfsExists(_fs, path.c_str()) == -1) {
+            derror_f("HDFS remove_path failed: path {} not found.", path);
+            resp.err = ERR_OBJECT_NOT_FOUND;
+            tsk->enqueue_with(resp);
+            return;
+        }
+
+        int entries = 0;
+        hdfsFileInfo *info = hdfsListDirectory(_fs, path.c_str(), &entries);
+        hdfsFreeFileInfo(info, entries);
+        if (entries > 0 && !req.recursive) {
+            derror_f("HDFS remove_path failed: directory {} is not empty.", path);
+            resp.err = ERR_DIR_NOT_EMPTY;
+            tsk->enqueue_with(resp);
+            return;
+        }
+
+        // Remove directory now.
+        if (hdfsDelete(_fs, path.c_str(), req.recursive) == -1) {
+            derror_f("HDFS remove_path {} failed.", path);
+            resp.err = ERR_FS_INTERNAL;
+        } else {
+            resp.err = ERR_OK;
+        }
+        tsk->enqueue_with(resp);
+    };
+
+    dsn::tasking::enqueue(LPC_HDFS_SERVICE_CALL, tracker, remove_path_background);
+    return tsk;
+}
+
+hdfs_file_object::hdfs_file_object(hdfs_service *s, const std::string &name)
+    : block_file(name), _service(s), _md5sum(""), _size(0), _has_meta_synced(false)
+{
+}
+
+error_code hdfs_file_object::get_file_meta()
+{
+    if (hdfsExists(_service->get_fs(), file_name().c_str()) == -1) {
+        dwarn_f("HDFS file {} does not exist.", file_name());
+        return ERR_OBJECT_NOT_FOUND;
+    }
+    hdfsFileInfo *info = hdfsGetPathInfo(_service->get_fs(), file_name().c_str());
+    if (info == nullptr) {
+        derror_f("HDFS get file info failed, file: {}.", file_name());
+        return ERR_FS_INTERNAL;
+    }
+    _size = info->mSize;
+    _has_meta_synced = true;
+    hdfsFreeFileInfo(info, 1);
+    return ERR_OK;
+}
+
+hdfs_file_object::~hdfs_file_object() {}
+
+error_code hdfs_file_object::write_data_in_batches(const char *data,
+                                                   const uint64_t data_size,
+                                                   uint64_t &written_size)
+{
+    written_size = 0;
+    hdfsFile writeFile =
+        hdfsOpenFile(_service->get_fs(), file_name().c_str(), O_WRONLY | O_CREAT, 0, 0, 0);
+    if (!writeFile) {
+        derror_f("Failed to open hdfs file {} for writting, error: {}.",
+                 file_name(),
+                 utils::safe_strerror(errno));
+        return ERR_FS_INTERNAL;
+    }
+    uint64_t cur_pos = 0;
+    uint64_t write_len = 0;
+    while (cur_pos < data_size) {
+        write_len = std::min(data_size - cur_pos, FLAGS_hdfs_write_batch_size_bytes);
+        tSize num_written_bytes = hdfsWrite(
+            _service->get_fs(), writeFile, (void *)(data + cur_pos), static_cast<tSize>(write_len));
+        if (num_written_bytes == -1) {
+            derror_f("Failed to write hdfs file {}, error: {}.",
+                     file_name(),
+                     utils::safe_strerror(errno));
+            hdfsCloseFile(_service->get_fs(), writeFile);
```

Comment:
Should we remove the file when error occurs?

Reply:
If writing/uploading file failed, we'll retry to overwrite the same file, so I think removing the file is not necessary.